esp/services/ws_machine/ws_machineService.cpp

```diff
@@ -2804,31 +2812,56 @@ void Cws_machineEx::readComponentUsageResult(IEspContext& context, IPropertyTree
     }
 }
 
+StringBuffer& Cws_machineEx::setUsageTimeStr(CUsageCache* usageCache, StringBuffer& timeStr)
+{
+    if (usageCache)
+        usageCache->queryTimeCached(timeStr);
+    else
+    {
+        CDateTime timeNow;
+        timeNow.setNow();
+        timeNow.getString(timeStr, true);
+    }
+    return timeStr;
+}
+
 bool Cws_machineEx::onGetComponentUsage(IEspContext& context, IEspGetComponentUsageRequest& req,
     IEspGetComponentUsageResponse& resp)
 {
     try
     {
         context.ensureFeatureAccess(FEATURE_URL, SecAccess_Read, ECLWATCH_MACHINE_INFO_ACCESS_DENIED, "Failed to Get Machine Information. Permission denied.");
 
-        StringBuffer cacheID;
-        buildComponentUsageCacheID(cacheID, req.getComponents());
-
-        if (!req.getBypassCachedResult() && readComponentUsageCache(context, cacheID, resp))
-            return true;
-
-        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(true);
-        Owned<IConstEnvironment> constEnv = envFactory->openEnvironment();
+        double version = context.getClientVersion();
 
+        StringBuffer timeStr;
+        IArrayOf<IEspComponentUsage> componentUsages;
         Owned<IPropertyTree> usageReq = createPTree("Req");
-        Owned<IPropertyTree> uniqueUsages = createPTree("Usage");
-        readComponentUsageReq(req, constEnv, usageReq, uniqueUsages);
+        if (!req.getBypassCachedResult())
```

Comment:
I will common up the code here. 

I am concerned the performance if forcing synchronous. For example, if the system has more than 400 machines (ex. an HPCC with 400 thor nodes), an ESP request with the BypassCachedResult flag is only interested in the usage of 1 machine.

Reply:
Why would going via the cache (and using latest info. from cache) be slow?
The only time it should block is if it has never built it (?)

Reply:
Talking to @wangkx - this may be something to do with a requirement that clients get immediate current state (of a machine not the whole cluster).

By passing the cache (which is designed to prevent waste and blockage) could mean clients degrade the performance of themselves and esp.
Not sure what use cases there are where the cache refresh time is inadequate for a client, if this usage cache is mainly for disk and cpu - and if it's good enough for Eclwatch page, it should be good enough for clients.

Reply:
I think this code could be simplified to make it easier to see what it's doing if you combined the code that is testing the bypass flag, and also had readComponentUsageReq create and return 'usageReq'.
Something like:

```
        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(true);
        Owned<IConstEnvironment> constEnv = envFactory->openEnvironment();
        Owned<IPropertyTree> usageReq = readComponentUsageReq(req, constEnv);

        Owned<CUsageCache> usage;
        Owned<IPropertyTree> uniqueUsages;
        if (req.getBypassCachedResult())
        {
            uniqueUsages.setown(createPTree("Usage"));
            //Add unique machines from the usageReq to uniqueUsages.
            setUniqueMachineUsageReq(usageReq, uniqueUsages);

            getMachineUsages(context, uniqueUsages);
        }
        else
        {
            usage.setown((CUsageCache*) usageCacheReader->getCachedInfo());
            if (!usage)
                throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Failed to get usage. Please try later.");
        }
```

Reply:
Will do